Bug 1820083: configure metal3 metrics collection

dhellmann opened the pull request:
* Refactor some of the proxy container setup so it can be reused.
* Add a metal3-metrics Service
* Add a metal3 ServiceMonitor

dhellmann commented:
/cc @JoelSpeed 

I don't think this is working right, but I'm putting it up for review so I can get some help understanding why.

dhellmann commented:
/cc @sadasu @honza

JoelSpeed commented on pkg/operator/baremetal_pod.go:
```diff
@@ -37,6 +39,35 @@ var volumeMounts = []corev1.VolumeMount{
+  namespaceSelector:
+    matchNames:
+      - openshift-machine-api
+  selector:
+    matchLabels:
+      k8s-app: controller
```
Is this not going to select the Machine API core components as well? Is there a specific label that is applied to only the metal3 components?

dhellmann replied:
There is not, right now. I'm exploring whether I can change that as part of this PR in some local changes I haven't published yet. I tried with an earlier version and got an error and undid the changes so I could make progress in other areas.

dhellmann replied:
It turns out that field is immutable, which means we have to delete and recreate the deployment to change it (at least as far as I can tell). It doesn't seem to hurt anything to have it this way for now, and we can do something about renaming when we move all of the metal3 logic into the CBO if it turns into a problem.

JoelSpeed replied:
My concern is that it is going to try and scrape all of the Machine API controllers, not just the metal3 ones. When you tried this out, did you notice in prometheus if it was trying to scrape more than just the one pod, it might have left some errors in the prometheus targets page showing that it couldn't scrape the machine-api-controllers pod?

dhellmann replied:
This new stuff will only be set up when the platform is "baremetal" though, right? So it will look at the cluster-api-provider-baremetal and metal3. Maybe there are other pods?

If we have to change the label, then I think we're going to have to defer fixing the bug to 4.7 or later, because we're trying to focus on building the new operator to manage metal3 and can't afford the time to build the logic to recreate the deployment with the new setting in the MAO.

enxebre replied:
Which field you mean immutable exactly Doug? This selector is for the new service you are defining in this PR https://github.com/openshift/machine-api-operator/pull/671/files#diff-03a356cadfb070910ca0faa53a4ab4bfR313thttps://github.com/openshift/machine-api-operator/pull/671/files#diff-03a356cadfb070910ca0faa53a4ab4bfR313 so you can choose anything.

dhellmann replied:
It is not possible to change the `k8s-app` label value in the Deployment.Spec.Selector.MatchLabels field. Attempting to change that with a patch like https://paste.centos.org/view/b0f842e8 causes the MAO to log an error every time it tries to update the Deployment:

```
W0916 15:42:54.324744 1050505 recorder_logging.go:46] &Event{ObjectMeta:{dummy.16355b11bbdfeaac  dummy    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},
InvolvedObject:ObjectReference{Kind:Pod,Namespace:dummy,Name:dummy,UID:,APIVersion:v1,ResourceVersion:,FieldPath:,},
Reason:DeploymentUpdateFailed,
Message:Failed to update Deployment.apps/metal3 -n openshift-machine-api: Deployment.apps "metal3" is invalid: spec.selector: Invalid value: 
v1.LabelSelector{MatchLabels:map[string]string{"api":"clusterapi", "k8s-app":"metal3"}, 
MatchExpressions:[]v1.LabelSelectorRequirement(nil)} field is immutable,
Source:EventSource{Component:,Host:,}, 
FirstTimestamp:2020-09-16 15:42:54.324665004 -0400 EDT m=+105.798620314,
LastTimestamp:2020-09-16 15:42:54.324665004 -0400 EDT m=+105.798620314,
Count:1,Type:Warning,
EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,}
```

I suspect that means adding a different label is going to result in the same error, because it would mean changing the same field in the Deployment.

zaneb replied:
https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#label-selector-updates

"In API version apps/v1, a Deployment's label selector is immutable after it gets created."

So yeah. This feels pretty wrong though, as requests will go to a bunch of things we aren't intending.

dhellmann replied:
They don't, though. The label selector is the same, but the *port* is not, and both are used to filter. Only the metal3 deployment matches both.

JoelSpeed commented on pkg/operator/baremetal_pod.go:
```diff
@@ -113,6 +144,44 @@ func createMariadbPasswordSecret(client coreclientv1.SecretsGetter, config *Oper
+					},
+				},
+			},
+			Selector: map[string]string{
+				"k8s-app": "controller",
+			},
```
Is this not going to select the Machine API core components as well? Is there a specific label that is applied to only the metal3 components?

JoelSpeed replied:
Same concern as above, this is going to try and expose the port on the `machine-api-onctrollers` pod as well isn't it? Or will this only come into effect when they're running in their own namespace?

If you check the endpoints for this service, which pods has it targeted?

dhellmann replied:
I see it includes the `machine-api-controllers` pod as well as the `metal3` pod. Those both run in the `openshift-machine-api` namespace.

JoelSpeed replied:
In that case, we need to verify that prometheus and in particular alerts are happy. IIRC there's an alert that first if any servicemonitor target is down, I would expect `machine-api-controllers` to show as down in this case since it won't have the correct port open?

enxebre replied:
This can may be match a specific label that should be added/replaced in the baremetal deployment so only those endpoints are covered by the service.

JoelSpeed commented:
@dhellmann Can you elaborate on why you don't think it's working, from the code there's nothing obvious that screams out that it shouldn't be working.

dhellmann commented:
> @dhellmann Can you elaborate on why you don't think it's working, from the code there's nothing obvious that screams out that it shouldn't be working.

Well, when I replace the MAO in my cluster with this code prometheus doesn't report any of the metrics I expect it to report. I'm currently rebuilding the cluster to try that again.

JoelSpeed commented:
> Well, when I replace the MAO in my cluster with this code prometheus doesn't report any of the metrics I expect it to report. I'm currently rebuilding the cluster to try that again.

Haha ok, I guess we just have to start debugging from the ground up on this one to find where the fault is. I'm trying to bring up a cluster-bot cluster with these changes, if that works I'll play around with that and see if I can advise. Otherwise I don't have access to a metal cluster to try this on.

dhellmann commented:
I've rebased the PR and re-tested locally and it seems to be working fine.

dhellmann commented:
/retest

JoelSpeed commented on pkg/metrics/metrics.go:
```diff
@@ -14,6 +14,7 @@ const (
 	DefaultHealthCheckMetricsAddress = ":8083"
 	DefaultMachineSetMetricsAddress  = ":8082"
 	DefaultMachineMetricsAddress     = ":8081"
+	DefaultMetal3MetricsAddress      = ":60000"
```
Would be good to know the motivation behind choosing this number, any reason not to try `8085`? (To match the exposed port)

JoelSpeed reviewed (commented):
Still concerned about this conflicting with the other pods in the namespace, any idea if there were issues when testing this?

sadasu reviewed (commented):
If we changed the label on the metal3 deployment object https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/baremetal_pod.go#L280 to something that doesn't clash with other deployments in MAO and update the metal3ServiceMonitorDefinition to look for this label, would that address concerns regarding scrapping the machine-api-controllers pod?

dhellmann commented:
> If we changed the label on the metal3 deployment object https://github.com/openshift/machine-api-operator/blob/master/pkg/operator/baremetal_pod.go#L280 to something that doesn't clash with other deployments in MAO and update the metal3ServiceMonitorDefinition to look for this label, would that address concerns regarding scrapping the machine-api-controllers pod?

It would, but that field is immutable. So we would need enough logic in the MAO to delete the Deployment and recreate it when the label values are wrong. See https://github.com/openshift/machine-api-operator/pull/671#discussion_r489712615

dhellmann commented on pkg/operator/baremetal_pod.go:
```diff
@@ -73,6 +75,35 @@ var volumes = []corev1.Volume{
+    matchLabels:
+      k8s-app: controller
+  endpoints:
+  - port: metrics
```
metal3-mtrc

dhellmann commented on pkg/operator/baremetal_pod.go:
```diff
@@ -264,6 +294,44 @@ func checkForBaremetalClusterOperator(osClient osclientset.Interface) (bool, err
+			Type: corev1.ServiceTypeNodePort,
+			Ports: []corev1.ServicePort{
+				{
+					Name: "metrics",
```
metal3-mtrc

dhellmann commented on pkg/operator/baremetal_pod.go:
```diff
@@ -264,6 +294,44 @@ func checkForBaremetalClusterOperator(osClient osclientset.Interface) (bool, err
+					Port: metal3ExposeMetricsPort,
+					TargetPort: intstr.IntOrString{
+						Type:   intstr.String,
+						StrVal: "metrics",
```
metal3-mtrc

dhellmann commented on pkg/operator/baremetal_pod.go:
```diff
@@ -395,65 +470,62 @@ func createInitContainerStaticIpSet(config *OperatorConfig, baremetalProvisionin
+
+func newMetal3Containers(config *OperatorConfig, baremetalProvisioningConfig BaremetalProvisioningConfig) []corev1.Container {
+	containers := []corev1.Container{
+		newKubeProxyContainer(config.Controllers.KubeRBACProxy, "metrics",
```
metal3-mtrc

dhellmann commented:
I have renamed the port used "metal3-mtrc" for consistency with other metrics collection ports and to ensure it is unique. I've also rebased the original set of patches.

JoelSpeed reviewed (approved):
/approve

Doug and I had a look at a cluster running this patch and found that prometheus also takes in the port name when deciding on targets. Because only the metal3 pod has the correct metric name, this is the only one that is being scraped.

Therefore, even though the pod labels overlap, there are no alerts or issues on the prometheus side.

JoelSpeed commented:
/retest

JoelSpeed commented:
/test e2e-metal-ipi

dhellmann commented:
/retest

dhellmann commented:
/retest

dhellmann commented:
/retest

dhellmann commented:
/retest

dhellmann commented:
/retest

JoelSpeed commented on install/0000_30_machine-api-operator_09_rbac.yaml:
```diff
@@ -323,6 +323,17 @@ rules:
       - list
       - patch
 
+  - apiGroups:
+      - ""
+    resources:
+      - services
+    verbs:
+      - create
+      - watch
+      - get
+      - list
+      - patch
+
```
Are we creating Services across namespaces? Should this be in the `Role` rather than the `ClusterRole`? Curious that we don't already have this, I thought we were creating a service from MAO already 🤔

dhellmann replied:
I didn't find any cases where the MAO was creating services from go code. There's one in a manifest file.

I've moved this block up to the Role instead of the ClusterRole because the new service is just created in the openshift-machine-api namespace.

dhellmann commented:
The previous failure was due to a panic when trying to apply the new ServiceMonitor. See https://github.com/openshift/library-go/pull/927 for a fix. I think the error is caused by not having RBAC set up. I have applied that fix to this PR to see if that's the error message. I will remove that fix and add the RBAC rules in another draft after getting the test results.

dhellmann commented:
/test e2e-metal-ipi

JoelSpeed commented:
/retest

dhellmann commented:
/retest

dhellmann commented:
/test e2e-metal-ipi

dhellmann commented:
/test e2e-metal-ipi

dhellmann commented:
The latest error is the baremetal-operator reporting that the provisioner is not ready because Ironic returns a 401 when it is asked whether it has loaded its drivers. I've rebased in case there was a configuration change related to auth that was missed in my branch.

zaneb commented on pkg/operator/baremetal_pod.go:
```diff
@@ -395,65 +470,62 @@ func createInitContainerStaticIpSet(config *OperatorConfig, baremetalProvisionin
+		Command: []string{"/baremetal-operator",
+			"-metrics-addr", metrics.DefaultMetal3MetricsAddress},
+		ImagePullPolicy: "IfNotPresent",
+		Env: []corev1.EnvVar{
```
Volume mounts are missing here, which is probably why it is failing to contact ironic.

dhellmann commented:
\o/ the metal job is passing finally!

/test e2e-gcp-operator

elmiko reviewed (commented):
this generally makes sense to me. i do have a question about the seelector issues raised by Joel, i couldn't quite tell if there was a resolution there. is this PR safe to merge with the expectation that those selector values will be revisited in the future to make the prometheus scrape more selective?

dhellmann commented:
> this generally makes sense to me. i do have a question about the seelector issues raised by Joel, i couldn't quite tell if there was a resolution there. is this PR safe to merge with the expectation that those selector values will be revisited in the future to make the prometheus scrape more selective?

The selectors ended up being selective enough to only pick up the single pod for metal3 because it used a combination of the label, which isn't unique, and the name of the ServicePort, which *is* unique.

elmiko commented:
> The selectors ended up being selective enough to only pick up the single pod for metal3 because it used a combination of the label, which isn't unique, and the name of the ServicePort, which _is_ unique.

perfect! thanks for the explanation

/lgtm

JoelSpeed commented:
/hold

@dhellmann This is going to need a rebase to remove `glog` in favour of `klog` as a PR just merged to upgrade from one to the other, the import for `klog` is `k8s.io/klog/v2` if you need it, should hopefully just be a case of updating it in these locations looking at the test failures:
```
# github.com/openshift/machine-api-operator/pkg/operator [github.com/openshift/machine-api-operator/pkg/operator.test]
pkg/operator/sync.go:329:3: undefined: glog
pkg/operator/sync.go:339:3: undefined: glog
# github.com/openshift/machine-api-operator/pkg/operator
pkg/operator/sync.go:329:3: undefined: glog
pkg/operator/sync.go:339:3: undefined: glog
```

dhellmann commented:
Rebased and `glog` removed. I guessed at the verbosity level based on the other calls in the same file, but let me know if I guessed wrong.

elmiko reviewed (approved):
i'm not 100% sure if `V(3)` is the most appropriate log level, but it is consistent with other usages.
/lgtm

dhellmann commented:
/retest

dhellmann commented:
/retest

@JoelSpeed the required CI jobs are passing and the Azure failures look like quota or other platform issues. I'm not sure what's up with GCP, but given that the job built a cluster and then a bunch of tests failed I don't expect that it's related to this PR. I'm retesting, but I think it's probably safe to remove your hold?

dhellmann commented:
I chatted with @elmiko about the hold on this, and he said it should be OK to remove it. I'm happy to rework anything in a follow-up, if that's necessary.

/hold cancel

elmiko commented:
as the code has been updated with regards to glog, i am removing the hold after talking with Doug.
/hold cancel

zaneb commented on pkg/operator/baremetal_pod.go:
```diff
@@ -264,6 +294,44 @@ func checkForBaremetalClusterOperator(osClient osclientset.Interface) (bool, err
+					Port: metal3ExposeMetricsPort,
+					TargetPort: intstr.IntOrString{
+						Type:   intstr.String,
+						StrVal: "metal3-mtrc",
```
This doesn't appear to be a real [IANA service type](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml). How does this work?
